### PR TITLE
Feat/input

### LINKED ISSUE #95: ### Input

### LINKED ISSUE #206: ### Autocomplete

### PR DESCRIPTION
**Pull request by hanx316**
- [x] Component Migration

#95

### CODE REVIEW

File: packages/input/src/index.vue [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+      _textareaCalcStyle.value = calcTextareaHeight(instance.refs.textarea, minRows, maxRows)
+    }
+    const setNativeInputValue = () => {
+      const input = getInput()
+      if (!input) return
+      if (input.value === nativeInputValue.value) return
+      input.value = nativeInputValue.value
+    }
+    const calcIconOffset = place => {
+      let elList = [].slice.call(instance.vnode.el.querySelectorAll(`.el-input__${place}`) || [])
```
Comment (SimonaliaChen) [inline packages/input/src/index.vue:311]: There are twice references to `instance.vnode.el`, maybe extract it out would be better

**Review by SimonaliaChen — CHANGES_REQUESTED** [review]

File: packages/input/src/index.vue [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+    <template v-if="type !== 'textarea'">
+      <!-- 前置元素 -->
+      <div v-if="$slots.prepend" class="el-input-group__prepend">
+        <slot name="prepend"></slot>
+      </div>
+      <input
+        v-if="type !== 'textarea'"
+        ref="input"
+        class="el-input__inner"
+        v-bind="$attrs"
```
Comment (Ryan2128) [inline packages/input/src/index.vue:28]: Hello, the `$attrs` maybe make a error, its behavior in below.

Incorrect behavior:
image

Correct behavior:
image


The `$attrs` pass the `class` attr of `el-input` component to `input`. `style` attr will be same behavior.

If delete this line directly, the `placeholder` will also not be passed.

Delete this line:
image

Comment (Ryan2128) [inline packages/input/src/index.vue:28]: Hello, the `$attrs` maybe make a error, its behavior in below.

Incorrect behavior:
image

Correct behavior:
image


The `$attrs` pass the `class` attr of `el-input` component to `input`. `style` attr will be same behavior.

If delete this line directly, the `placeholder` will also not be passed.

Delete this line:
image

**Review by Ryan2128 — CHANGES_REQUESTED** [review]

File: packages/input/src/index.vue (hunk 2 above)
Comment (Ryan2128) [reply packages/input/src/index.vue:28]: In vue3, the `class` and `style` are included  in `$attrs`.

File: packages/input/src/index.vue (hunk 1 above)
Comment (JeremyWuuuuu) [inline packages/input/src/index.vue:311]: Use `Array.from()`

File: packages/input/src/index.vue [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+    const handleInput = event => {
+      // should not emit input during composition
+      // see: https://github.com/ElemeFE/element/issues/10516
+      if (isComposing.value) return
+
+      // hack for https://github.com/ElemeFE/element/issues/8548
+      // should remove the following line when we don't support IE
+      if (event.target.value === nativeInputValue.value) return
+
+      ctx.emit('update:modelValue', event.target.value)
```
Comment (JeremyWuuuuu) [inline packages/input/src/index.vue:346]: use `UPDATE_MODEL_EVENT` in `@utils/constants`

**Review by jw-foss — CHANGES_REQUESTED** [review]
One other suggestion here, move related logics into a single file here might help with the readability

File: packages/input/src/index.vue [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+      validator: (val: string): boolean => ['', 'none', 'both', 'horizontal', 'vertical'].indexOf(val) > -1,
+    },
+    autosize: {
+      type: [Boolean, Object],
+      default: false,
+    },
+    autocomplete: {
+      type: String,
+      default: 'off',
+      validator: (val: string): boolean => ['on', 'off'].indexOf(val) > -1,
```
[the remark this answers is not in the document]
Comment (hanx316) [reply packages/input/src/index.vue:165]: seems like the env supports `Proxy` must support `Array.prototype.includes`...i'll replace it.

File: packages/input/src/index.vue (hunk 2 above)
Comment (SimonaliaChen) [reply packages/input/src/index.vue:28]: It'll be optimized in future releases according to #235

**Review by jw-foss — APPROVED** [review]
LGTM

**Review by zazzaz — APPROVED** [review]
L Cool TM

**Review by SimonaliaChen — APPROVED** [review]

**Pull request merged**